realm/realm-library/src/main/java/io/realm/ObjectChangeSet.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2017 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm;
+
+/**
+ * Information about the changes made to an object.
+ * @see RealmObject#addChangeListener(RealmObjectChangeListener) .
+ */
+public interface ObjectChangeSet {
+
+    /**
+     * Information about a specific field which changed in an {@link RealmObject} change notification.
+     */
+    class FieldChange {
+        /**
+         * The name of the field which changed.
+         */
+        final String fieldName;
+
+        public FieldChange(String fieldName) {
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * @return {@code null} if the object has been deleted. Otherwise returns the information about changed fields.
+     */
+    FieldChange[] getFieldChanges();
```

Comment:
1. I prefer one listener on the `RealmObject` instead of multiple listeners on the object's fields. It makes the API simpler and it is still quite simple for real use cases, eg.: databinding. We/Users can always inherit the `RealmObjectChangeListener` for specific use cases. See my example code of using it in data binding https://github.com/realm/realm-android-adapters/pull/99/files#diff-58a0e399a032fe72057d0e65269ebdb6

2. I would choose something with less methods, so probably:

```java
public interface ObjectChangeset {
  boolean isDeleted();
  String[] getChangedField();
 
  public static class FieldChange<T> {
    T oldValue;
    T newValue;
  }

  // This will always return null for RealmList field.
  <E> FieldChange<E> getChange(E type, String fieldName);
}
```

3. `FieldChange` and `getChange()` will not be implemented in this PR. Also if we are going to have them, there are some restrictions:

a) For local transactions, there is no way to get the old values.
b) store the old values will be slow and consuming memory when the old value is a big String or Binary -- The listener is on the `RealmObject` instead of on the field.